File: backend/app/schemas/artifact.py
"""Artifact Pydantic Schema — 伪影生成/分类/修复请求与响应"""

from typing import Any, Dict, List, Literal, Optional
from pydantic import BaseModel, Field


class ArtifactGenerateRequest(BaseModel):
    """伪影生成请求"""
    source: Literal["atlas", "procedural", "dicom"] = "atlas"
    case_id: Optional[str] = None
    series_id: Optional[str] = None
    size: int = Field(160, ge=64, le=192)
    artifact_types: List[str]
    artifact_params: Dict[str, Dict[str, Any]] = {}
    ct_params: Optional[Dict[str, Any]] = None


class ArtifactGenerateResponse(BaseModel):
    """伪影生成响应"""
    artifact_volume_base64: str
    clean_volume_base64: str
    artifact_masks_base64: Dict[str, str]
    metadata: Dict[str, Any]
    standardized_case: Dict[str, Any]


class ArtifactClassifyRequest(BaseModel):
    """伪影分类请求"""
    source: str = "phantom"
    series_id: Optional[str] = None
    slice_indices: Optional[List[int]] = None


class SliceClassifyResult(BaseModel):
    scores: Dict[str, float]
    labels: List[str]
    dominant: str
    slice_index: int


class ArtifactClassifyResponse(BaseModel):
    """伪影分类响应"""
    overall_scores: Dict[str, float]
    per_slice_scores: List[SliceClassifyResult]
    dominant_artifact: str
    slice_count: int


class ArtifactRestoreRequest(BaseModel):
    """伪影修复请求"""
    source: str = "phantom"
    series_id: Optional[str] = None
    artifact_types: List[str] = Field(default_factory=list, description="已知伪影类型列表")
    method: Literal["traditional", "deep_learning", "auto"] = "auto"
    shape: Optional[List[int]] = None
    spacing: Optional[List[float]] = None


class ArtifactRestoreResponse(BaseModel):
    """伪影修复响应"""
    restored_slice: List[List[float]]
    original_slice: List[List[float]]
    quality_metrics: Dict[str, float]
    steps: List[Dict[str, Any]]
    report: str
    shape: List[int]
    spacing: List[float]


class ArtifactPipelineRequest(BaseModel):
    """完整流水线: 生成 → 分类 → 修复"""
    source: str = "phantom"
    series_id: Optional[str] = None
    artifact_type: str = "noise"
    params: Dict[str, Any] = Field(default_factory=dict)
    slice_index: Optional[int] = None


class ArtifactPipelineResponse(BaseModel):
    """完整流水线响应"""
    artifact_type: str
    classification: Dict[str, Any]
    restoration: Dict[str, Any]
    original_slice: List[List[float]]
    artifact_slice: List[List[float]]
    restored_slice: List[List[float]]
    quality_metrics: Dict[str, float]
    metadata: Dict[str, Any]


class ArtifactJobResponse(BaseModel):
    """作业响应"""
    id: str
    job_type: str
    status: str
    config: Optional[Dict[str, Any]]
    output_path: Optional[str]
    quality_metrics: Optional[Dict[str, Any]]
    created_at: Optional[str]
    completed_at: Optional[str]


class ArtifactTypesResponse(BaseModel):
    types: List[str]


class ArtifactGenerateSliceRequest(BaseModel):
    """单切片生成请求 (兼容现有前端)"""
    artifact_type: str
    params: Dict[str, Any] = Field(default_factory=dict)
    source: str = "phantom"
    series_id: Optional[str] = None
    study_id: Optional[str] = None
    slice_index: Optional[int] = None


class ArtifactGenerateSliceResponse(BaseModel):
    """单切片生成响应 (兼容现有前端)"""
    artifact_type: str
    original_slice: List[List[float]]
    artifact_slice: List[List[float]]
    mask_slice: List[List[float]]
    metadata: Dict[str, Any]
    shape: List[int]
    spacing: List[float]
    source: str

File: frontend/src/types/artifact.ts
/** 伪影类型 */
export type ArtifactType =
  | 'metal'
  | 'motion'
  | 'noise'
  | 'ring'
  | 'streak'
  | 'beam_hardening'
  | 'composite';

/** 伪影生成参数 */
export interface ArtifactParams {
  metal?: MetalArtifactParams;
  motion?: MotionArtifactParams;
  noise?: NoiseArtifactParams;
  ring?: RingArtifactParams;
  streak?: StreakArtifactParams;
  beamHardening?: BeamHardeningParams;
}

export interface MetalArtifactParams {
  metalType: 'titanium' | 'stainless_steel' | 'dental_amalgam' | 'gold' | 'copper';
  center: [number, number, number];
  radiusMm: [number, number, number];
  streakIntensity: number;
  beamHardeningStrength: number;
}

export interface MotionArtifactParams {
  motionType: 'respiratory' | 'cardiac' | 'random';
  amplitudeMm: number;
  blurSigma: number;
}

export interface NoiseArtifactParams {
  mAs: number;
  referenceMAs: number;
}

export interface RingArtifactParams {
  numRings: number;
  intensity: number;
}

export interface StreakArtifactParams {
  numStreaks: number;
  intensity: number;
}

export interface BeamHardeningParams {
  cuppingStrength: number;
  darkBandStrength: number;
}

/** 伪影生成请求/响应 */
export interface ArtifactGenerateRequest {
  source: 'phantom' | 'dicom';
  seriesId?: string;
  artifactType: string;
  params: Record<string, unknown>;
  sliceIndex?: number;
}

export interface ArtifactGenerateResponse {
  artifactType: string;
  originalSlice: number[][];
  artifactSlice: number[][];
  maskSlice: number[][];
  metadata: Record<string, unknown>;
  shape: number[];
  spacing: number[];
  source: string;
}

/** 分类请求 */
export interface ClassifyRequest {
  source?: 'phantom' | 'dicom';
  seriesId?: string;
  sliceIndices?: number[];
}

/** 分类响应 (alias) */
export type ClassifyResponse = ArtifactClassificationResult;

/** 伪影分类结果 */
export interface SliceClassifyResult {
  scores: Record<string, number>;
  labels: string[];
  dominant: string;
  sliceIndex: number;
}

export interface ArtifactClassificationResult {
  overallScores: Record<string, number>;
  perSliceScores: SliceClassifyResult[];
  dominantArtifact: string;
  sliceCount: number;
}

/** 伪影修复结果 */
export interface ArtifactRestoreResponse {
  restoredSlice: number[][];
  originalSlice: number[][];
  qualityMetrics: QualityMetrics;
  steps: RestoreStep[];
  report: string;
  shape: number[];
  spacing: number[];
}

export interface QualityMetrics {
  psnr: number;
  ssim: number;
  nmse: number;
  maeHu: number;
  psnrImprovement?: number;
  ssimImprovement?: number;
}

export interface RestoreStep {
  artifactType: string;
  strategy: string;
  psnrDelta: number;
}

/** 完整流水线结果 */
export interface ArtifactPipelineResponse {
  artifactType: string;
  classification: Record<string, unknown>;
  restoration: { steps: RestoreStep[]; report: string };
  originalSlice: number[][];
  artifactSlice: number[][];
  restoredSlice: number[][];
  qualityMetrics: QualityMetrics;
  metadata: Record<string, unknown>;
}

/** 训练请求/响应 */
export interface TrainRequest {
  epochs: number;
  batchSize: number;
  learningRate: number;
  numVolumes: number;
  outputDir: string;
}

export interface TrainEpochResult {
  epoch: number;
  trainLoss: number;
  valLoss: number;
  trainAcc: number;
  valAcc: number;
  trainF1: number;
  valF1: number;
}

export interface TrainStatusResponse {
  status: string;
  currentEpoch: number;
  totalEpochs: number;
  trainLoss: number;
  valLoss: number;
  trainAcc: number;
  valAcc: number;
  trainF1: number;
  valF1: number;
  bestValLoss: number;
  epochHistory: TrainEpochResult[];
  error: string | null;
  startTime: number | null;
}

export interface TrainHistoryResponse {
  epochs: TrainEpochResult[];
  bestValLoss: number;
  outputDir: string;
}

/** CT 序列信息 */
export interface SeriesInfo {
  id: string;
  studyId: string;
  description: string | null;
  modality: string | null;
  imageCount: number | null;
  rows: number | null;
  columns: number | null;
}

/** 作业信息 */
export interface ArtifactJob {
  id: string;
  jobType: string;
  status: string;
  config: Record<string, unknown> | null;
  outputPath: string | null;
  qualityMetrics: Record<string, unknown> | null;
  createdAt: string | null;
  completedAt: string | null;
}
